File: frontend/src/api/postWorks.ts
import { apiClient } from '@/api/client';
import type { PaginatedResponse } from '@/types';

export interface PostWorksState {
  id: string;
  building_id: string;
  intervention_id: string | null;
  state_type: string;
  pollutant_type: string | null;
  title: string;
  description: string | null;
  zone_id: string | null;
  element_id: string | null;
  material_id: string | null;
  verified: boolean;
  verified_by: string | null;
  verified_at: string | null;
  evidence_json: string | null;
  recorded_by: string | null;
  recorded_at: string;
  notes: string | null;
}

export interface BeforeAfterComparison {
  building_id: string;
  intervention_id: string | null;
  before: {
    total_positive_samples: number;
    by_pollutant: Record<string, number>;
    risk_areas: Array<{ pollutant: string; location: string; risk_level: string }>;
  };
  after: {
    removed: number;
    remaining: number;
    encapsulated: number;
    treated: number;
    unknown_after_intervention: number;
    recheck_needed: number;
    by_pollutant: Record<string, Record<string, number>>;
  };
  summary: {
    remediation_rate: number;
    verification_rate: number;
    residual_risk_count: number;
  };
}

export interface PostWorksSummary {
  building_id: string;
  total_states: number;
  by_state_type: Record<string, number>;
  by_pollutant: Record<string, number>;
  verification_progress: {
    verified: number;
    unverified: number;
    rate: number;
  };
  interventions_covered: number;
}

export const postWorksApi = {
  list: async (buildingId: string): Promise<PaginatedResponse<PostWorksState>> => {
    const response = await apiClient.get<PaginatedResponse<PostWorksState>>(`/buildings/${buildingId}/post-works`, {
      params: { size: 50 },
    });
    return response.data;
  },
  compare: async (buildingId: string, interventionId?: string): Promise<BeforeAfterComparison> => {
    const params: Record<string, string> = {};
    if (interventionId) params.intervention_id = interventionId;
    const response = await apiClient.get<BeforeAfterComparison>(`/buildings/${buildingId}/post-works/compare`, {
      params,
    });
    return response.data;
  },
  summary: async (buildingId: string): Promise<PostWorksSummary> => {
    const response = await apiClient.get<PostWorksSummary>(`/buildings/${buildingId}/post-works/summary`);
    return response.data;
  },
  verify: async (buildingId: string, stateId: string, notes?: string): Promise<PostWorksState> => {
    const response = await apiClient.put<PostWorksState>(`/buildings/${buildingId}/post-works/${stateId}`, {
      verified: true,
      notes,
    });
    return response.data;
  },
};

File: backend/app/schemas/post_works.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class PostWorksStateCreate(BaseModel):
    intervention_id: uuid.UUID | None = None
    state_type: str
    pollutant_type: str | None = None
    title: str
    description: str | None = None
    zone_id: uuid.UUID | None = None
    element_id: uuid.UUID | None = None
    material_id: uuid.UUID | None = None
    verified: bool = False
    evidence_json: str | None = None
    notes: str | None = None


class PostWorksStateUpdate(BaseModel):
    intervention_id: uuid.UUID | None = None
    state_type: str | None = None
    pollutant_type: str | None = None
    title: str | None = None
    description: str | None = None
    zone_id: uuid.UUID | None = None
    element_id: uuid.UUID | None = None
    material_id: uuid.UUID | None = None
    verified: bool | None = None
    verified_by: uuid.UUID | None = None
    verified_at: datetime | None = None
    evidence_json: str | None = None
    notes: str | None = None


class PostWorksStateRead(BaseModel):
    id: uuid.UUID
    building_id: uuid.UUID
    intervention_id: uuid.UUID | None
    state_type: str
    pollutant_type: str | None
    title: str
    description: str | None
    zone_id: uuid.UUID | None
    element_id: uuid.UUID | None
    material_id: uuid.UUID | None
    verified: bool
    verified_by: uuid.UUID | None
    verified_at: datetime | None
    evidence_json: str | None
    recorded_by: uuid.UUID | None
    recorded_at: datetime
    notes: str | None

    model_config = ConfigDict(from_attributes=True)
